src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,220 @@
+                    throw new InvalidOperationException($"No extensions registered at '{assemblyFolderPath}'");
+            }
+
+            var assemblyHandlers = await extensionFolder.GetAssemblyHandlersAsync(assemblyFilePath, cancellationToken).ConfigureAwait(false);
+
+            return new(
+                [.. assemblyHandlers.WorkspaceMessageHandlers.Keys],
+                [.. assemblyHandlers.DocumentMessageHandlers.Keys]);
+        }
+
+        private ValueTask<VoidResult> ResetInCurrentProcessAsync()
+        {
+            lock (_gate)
+            {
+                _folderPathToExtensionFolder_useOnlyUnderLock.Clear();
+                ClearCachedHandlers_WhileUnderLock();
+                return default;
+            }
+        }
+
+        private async ValueTask<string> HandleExtensionMessageInCurrentProcessAsync<TArgument>(
+            TArgument executeArgument, bool isSolution, string messageName, string jsonMessage,
+            Dictionary<string, AsyncLazy<ImmutableArray<IExtensionMessageHandlerWrapper>>> cachedHandlers,
+            CancellationToken cancellationToken)
+        {
+            AsyncLazy<ImmutableArray<IExtensionMessageHandlerWrapper>> lazyHandlers;
+            lock (_gate)
+            {
+                // May be called a lot.  So we use the non-allocating form of this lookup pattern.
+                lazyHandlers = cachedHandlers.GetOrAdd(
+                    messageName,
+                    static (messageName, arg) => AsyncLazy.Create(
+                        static (arg, cancellationToken) => arg.@this.ComputeHandlersAsync(arg.messageName, arg.isSolution, cancellationToken),
+                        (messageName, arg.@this, arg.isSolution)),
+                    (messageName, @this: this, isSolution));
+            }
+
+            var handlers = await lazyHandlers.GetValueAsync(cancellationToken).ConfigureAwait(false);
+            if (handlers.Length == 0)
+                throw new InvalidOperationException($"No handler found for message {messageName}.");
+
+            if (handlers.Length > 1)
+                throw new InvalidOperationException($"Multiple handlers found for message {messageName}.");
+
+            var handler = handlers[0];
+            if (s_disabledExtensionHandlers.TryGetValue(handler, out _))
+                throw new InvalidOperationException($"Handler was disabled due to previous exception.");
+
+            try
+            {
+                var message = JsonSerializer.Deserialize(jsonMessage, handler.MessageType);
+                var result = await ((IExtensionMessageHandlerWrapper<TArgument>)handler).ExecuteAsync(
+                    message, executeArgument, cancellationToken).ConfigureAwait(false);
+                return JsonSerializer.Serialize(result, handler.ResponseType);
+            }
+            catch (Exception ex) when (DisableHandlerAndPropagate(ex))
```

Comment:
Few things
1.  Seems like cancellation exceptions need to be excluded (don't want to disable a handler if just cancelled)
2.  Might be a bit restrictive to disable a handler after a single exception?  Maybe we can have a counter or something?  Or x exceptions in a certain time (maybe followup)

Reply:
you're correct.  cancellation needs to be ok.

wrt to disabling, this is what we do with our other extension scenarios as well.  Though after disabling we do offer to reenable if the user wants to.  We can do that in a followup.

Reply:
We don't have the UI support for disabling / re-enabling in VSCode, so that is definitely something we'd want to add.  

Additionally, LSP sometimes uses exceptions to signal the request couldn't be handled, but not catastrophically so.  For example, sometimes we throw [content modified exceptions](https://github.com/dotnet/roslyn/blob/32c1fcf99c839dc9a0fe2f9d26e0d3b1257ad0d5/src/LanguageServer/Protocol/Handler/CodeLens/CodeLensResolveHandler.cs#L58) if the request is out of date.  So I'm not 100% certain that we should disable so aggressively.